src/cargo/util/toml/mod.rs

```diff
@@ -880,11 +964,74 @@ where
         where
             E: de::Error,
         {
-            string.trim().parse().map_err(de::Error::custom)
+            match string.trim().parse().map_err(de::Error::custom) {
+                Ok(parsed) => Ok(MaybeWorkspace::Defined(parsed)),
+                Err(e) => Err(e),
+            }
+        }
+
+        fn visit_map<V>(self, map: V) -> Result<Self::Value, V::Error>
+        where
+            V: de::MapAccess<'de>,
+        {
+            let mvd = de::value::MapAccessDeserializer::new(map);
+            TomlWorkspaceField::deserialize(mvd).map(MaybeWorkspace::Workspace)
+        }
+    }
+
+    deserializer.deserialize_any(Visitor)
+}
+
+/// Enum that allows for the parsing of { workspace = true } in a Cargo.toml
+///
+/// It allows for things to be inherited from a workspace or defined as needed
+#[derive(Deserialize, Serialize, Clone, Debug)]
+#[serde(untagged)]
+pub enum MaybeWorkspace<T> {
+    Workspace(TomlWorkspaceField),
+    Defined(T),
+}
+
+impl<T> MaybeWorkspace<T> {
+    fn resolve<'a>(
+        self,
+        cargo_features: &Features,
+        label: &str,
+        inherit: impl FnOnce() -> Option<&'a InheritableFields>,
+        get_ws: impl FnOnce(&InheritableFields) -> Option<T>,
+    ) -> CargoResult<T> {
+        match self {
+            MaybeWorkspace::Defined(value) => Ok(value),
+            MaybeWorkspace::Workspace(TomlWorkspaceField { workspace: true }) => {
+                cargo_features.require(Feature::workspace_inheritance())?;
+                let inherit = inherit().context(format!(
+                    "You cannot inherit fields from a parent workspace currently, tried to on {}",
```

Comment:
Might be good to say tell the user why.

Also, errors generally start with lowercase

Reply:
would "tried to inherit from a workspace for {}, but the workspace is not allowed to specify inheritable fields at this time" be better?

Reply:
Under what condition do we not have inheritable fields?  I'm looking at `deny_inherit_fields_from_parent_workspace` and `deny_inherit_dependencies_from_parent_workspace` and its not too clear to me.

Reply:
I'll amend my post above to better explain this but ill also describe it here. 

Part 1 has no tooling to allow inheriting from a workspace that is in a different `Cargo.toml`. This mean that you can only inherit from inside the same `Cargo.toml`, i.e. it has both a `[workspace]` and a `[package]`. This was done so that we could split the implementation of `Maybeworkspace<T>` and `Workspace(TomlWorkspaceDependency)` up from the searching for a workspace root and keeping the cache and various other things. By allowing inheriting from the same `Cargo.toml` only, it makes these changes testable without adding in the searching and other various things in the same that PR caused too much complexity in the previous PR.

Reply:
So it sounds like this error is an internal error about the feature not being implemented yet?  Seems like that is what we should have the error message focus on.  Something like "Inheriting from a parent workspace is not implemented yet".

I see how your original message and the follow up were intending to say that but there was just enough wiggle room for misinterpretation that I didn't pick this up.

Could you also clarify that with the test names?  Something like "error_on_unimplemented_inheritance()`.  This will make it clearer that the contract the test is enforcing is temporary rather than being fundamental.

Reply:
I'll change it to what you said. This could happen in Part 2 if you can't find a workspace root but I think we could just change the message in Part 2